### ISSUE TITLE
413 response does not call HttpErrorHandler.onClientError

### ISSUE THREAD

**Issue by mkurz** [playframework/playframework#4704]
Starting with Play 2.4 you get a `413 Request Entity Too Large` response when you exceed the maximum body length - see [the migration Guide](https://www.playframework.com/documentation/2.4.x/Migration24#Maximum-body-length).

The problem:
There is no way to handle which page or result to return/render.
Also `HttpErrorHandler.onClientError` is not called.

I consider this a bug - also because before we could call something like `if RequestBody.isMaxSizeExceeded...` (which now always returns `false`) and act accordingly - which means there is no upgrade path for this functionality.

**Comment by mkurz** [comment]
If it is possible I would love to have this fixed in 2.4.1, so we can finish our 2.4 migrations. Actually this is the last thing that bothers us.
If someone could quickly point me out where this has be to be fixed, I could try to set up a PR.

**Comment by richdougherty** [comment]
The place to fix it is in `ContentTypes.scala` where all the `BodyParser`s are defined. Look for occurrences of `Results.EntityTooLarge`, you'll want to change this hardcoded result to a call to `HttpErrorHandler.onClientError`. The `Iteratee.eofOrElse(Results.EntityTooLarge)`'s argument is by-name, so it should be OK to call `onClientError` as an argument to that method.

To get the current `HttpErrorHandler` you can use `LazyHttpErrorHandler` defined in `HttpErrorHandler.scala`. The `LazyHttpErrorHandler` looks up the handler from the current application every time it is called.

While you're at it, would you mind changing the `LazyHttpErrorHandler` to use `Application.instanceCache`, so it caches the current `HttpErrorHandler` rather than going through Guice every time? To use the cache, just change the line:

```scala
object LazyHttpErrorHandler extends HttpErrorHandler {
  private def ErrorHandler = Play.maybeApplication.fold[HttpErrorHandler](DefaultHttpErrorHandler)(_.errorHandler)
```

to

```scala
object LazyHttpErrorHandler extends HttpErrorHandler {
  private val errorHandlerCache = Application.instanceCache[HttpErrorHandler]
  private def ErrorHandler = Play.maybeApplication.fold[HttpErrorHandler](DefaultHttpErrorHandler)(errorHandlerCache)
```
